graphql/index.ts

```diff
@@ -4,15 +4,56 @@ import { UserModel } from '../models/UserProfile';
 
 const UserTC = composeWithMongoose(UserModel, {});
 
+function requireToken(resolvers) {
+  return Object.keys(resolvers).reduce((acc, key) => {
+    return {
+      ...acc,
+      [key]: resolvers[key].wrapResolve(next => rp => {
+        if (!rp.context.jwt.valid) {
+          throw new Error('You must provide valid token');
+        }
+        return next(rp);
+      }),
+    };
+  }, {});
+}
+
+function requireSameUserMongoId(resolvers) {
+  return Object.keys(resolvers).reduce((acc, key) => {
+    return {
+      ...acc,
+      [key]: resolvers[key].wrapResolve(next => async rp => {
+        const _id = rp.args._id || rp.args.record._id;
+        const egoId = await UserModel.findOne({ _id }).then(
+          user => user.ego_id,
+        );
+
+        if (rp.args.record && rp.args.record.ego_id !== egoId) {
+          throw new Error("You can't change your ego id");
+        } else if (`${egoId}` !== `${rp.context.jwt.payload.sub}`) {
+          throw new Error("You can't edit someone elses profile");
+        } else {
+          return next(rp);
+        }
+      }),
+    };
+  }, {});
+}
+
 GQC.rootQuery().addFields({
   user: UserTC.getResolver('findById'),
   users: UserTC.getResolver('pagination'),
 });
 
-GQC.rootMutation().addFields({
-  userCreate: UserTC.getResolver('createOne'),
-  userUpdate: UserTC.getResolver('updateById'),
-  userRemove: UserTC.getResolver('removeById'),
-});
+GQC.rootMutation().addFields(
+  requireToken({
+    userCreate: UserTC.getResolver('createOne'),
+
+    ...requireSameUserMongoId({
+      userRemove: UserTC.getResolver('removeById'),
+      userUpdate: UserTC.getResolver('updateById'),
+    }),
+  }),
+);
```

Comment:
I wonder if we could reference this -

```js
  restrictToAuthenticated: satisfy(isAuthenticated),
  restrictToAdmin: satisfy(isAdmin),
  restrictToSelf: satisfy([isSelf, isAdmin]),
  restrictToMembers: satisfy([isMember, isAdmin]),
```

where `satisfy` is 

```js
var satisfy = function (conditions, all) {
  if (!Array.isArray(conditions)) {
    conditions = [conditions];
  }
  return function(req, res, next) {
    var results = conditions.map(function (condition) {
      return condition(req);
    });
    if ((all ? _.all : _.any)(results)) {
      next();
    } else {
      res.status(401).send({message: 'unauthorized'});
    }
  };
};
```

Reply:
How do you feel about something that looks like this - 

```js
GQC.rootMutation().addFields({
  userCreate: restrict(UserTC.getResolver('createOne'), restrictToAuthenticated),
  userUpdate: restrict(UserTC.getResolver('updateById'), restrictToSelf),
  userRemove: restrict(UserTC.getResolver('removeById'), restrictToSelf),
});
```

which could accommodate this - 

```js
GQC.rootMutation().addFields({
  userCreate: restrict(UserTC.getResolver('createOne'), [restrictToAuthenticated, restrictToAdmin]),
  userUpdate: restrict(UserTC.getResolver('updateById'), [restrictToSelf, restrictToAdmin]),
  userRemove: restrict(UserTC.getResolver('removeById'), [restrictToSelf, restrictToAdmin]),
});
```

Where `restrict(resolver, [conditionA, conditionB])` would allow condition A or B, while `restrict(resolver, conditionA, conditionB)` would require conditions A in addition to B?

Overcomplicating things?

Reply:
I think thats good, made an update 👍